File: app/src/main/java/com/binatestation/kickstart/fragments/dialogs/MultiSelectDialogFragment.java
package com.binatestation.kickstart.fragments.dialogs;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.binatestation.kickstart.R;
import com.binatestation.kickstart.adapters.RecyclerViewAdapter;
import com.binatestation.kickstart.listeners.OnListItemClickListener;
import com.binatestation.kickstart.models.MultiSelectDataModel;

import java.util.ArrayList;

/**
 * Created by RKR on 11-09-2018.
 * MultiSelectDialogFragment.
 */
public class MultiSelectDialogFragment extends BaseDialogFragment implements OnListItemClickListener, View.OnClickListener {
    public static final String TAG = "MultiSelectDialogFragme";

    private static final String KEY_REQUEST_CODE = "REQUEST_CODE";
    private int mRequestCode;
    private boolean mMultiSelectable = true;
    private Object mSelectedItem;

    private RecyclerView mRecyclerView;
    private RecyclerViewAdapter mAdapter;
    private LinearLayoutManager mLinearLayoutManager;

    private MultiSelectListener mListener;

    public static MultiSelectDialogFragment newInstance(int requestCode) {

        Bundle args = new Bundle();
        args.putInt(KEY_REQUEST_CODE, requestCode);
        MultiSelectDialogFragment fragment = new MultiSelectDialogFragment();
        fragment.setArguments(args);
        fragment.setCancelable(false);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle bundle = getArguments();
        if (bundle != null) {
            mRequestCode = bundle.getInt(KEY_REQUEST_CODE);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_multi_select, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mRecyclerView = view.findViewById(R.id.recycler_view);

        mRecyclerView.setLayoutManager(mLinearLayoutManager = new LinearLayoutManager(getContext()));
        mRecyclerView.setAdapter(getAdapter());
        View actionPositive = view.findViewById(R.id.action_positive);
        View actionNegative = view.findViewById(R.id.action_negative);

        actionPositive.setOnClickListener(this);
        actionNegative.setOnClickListener(this);
    }

    public void setListener(MultiSelectListener mListener) {
        this.mListener = mListener;
    }

    /**
     * gets the instance of LinearLayoutManager
     *
     * @return LinearLayoutManager
     */
    public LinearLayoutManager getLinearLayoutManager() {
        return mLinearLayoutManager;
    }

    /**
     * gets the recyclerView
     *
     * @return RecyclerView
     */
    public RecyclerView getRecyclerView() {
        return mRecyclerView;
    }

    public void setMultiSelectable(boolean multiSelectable) {
        this.mMultiSelectable = multiSelectable;
    }

    /**
     * gets the adapter
     *
     * @return RecyclerViewAdapter
     */
    public RecyclerViewAdapter getAdapter() {
        if (mAdapter == null) {
            mAdapter = new RecyclerViewAdapter();
            mAdapter.setClickListener(this);
        }
        return mAdapter;
    }

    public void setData(ArrayList<?> data) {
        if (data != null) {
            getAdapter().setTypedData(MultiSelectDataModel.parseData(data));
        }
    }

    @Override
    public void onClickItem(Object object, int position, View actionView) {
        if (!mMultiSelectable) {
            mSelectedItem = object;
            actionPositive();
        }
    }

    public ArrayList<Object> getSelectedItems() {
        ArrayList<Object> selectedObjects = new ArrayList<>();
        if (mMultiSelectable) {
            ArrayList<?> objects = getAdapter().getData();
            for (Object object : objects) {
                if (object instanceof MultiSelectDataModel) {
                    MultiSelectDataModel multiSelectDataModel = (MultiSelectDataModel) object;
                    if (multiSelectDataModel.isSelected())
                        selectedObjects.add(multiSelectDataModel.getItem());
                }
            }
        } else {
            if (mSelectedItem instanceof MultiSelectDataModel) {
                MultiSelectDataModel selectedItem = (MultiSelectDataModel) mSelectedItem;
                selectedObjects.add(selectedItem.getItem());
            }
        }
        return selectedObjects;
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.action_positive) {
            actionPositive();
        } else if (v.getId() == R.id.action_negative) {
            dismiss();
        }
    }

    private void actionPositive() {
        if (mListener != null) {
            mListener.onSubmit(getSelectedItems(), mRequestCode);
        }
        dismiss();
    }

    public interface MultiSelectListener {
        void onSubmit(ArrayList<Object> objects, int requestCode);
    }
}

File: kickoff/src/main/java/com/binatestation/android/kickoff/utils/fragments/dialogs/MultiSelectDialogFragment.kt
/*
 * (c) Binate Station Private Limited. All rights reserved.
 */

@file:Suppress("unused")

package com.binatestation.android.kickoff.utils.fragments.dialogs

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.binatestation.android.kickoff.R
import com.binatestation.android.kickoff.repository.models.Selectable
import com.binatestation.android.kickoff.utils.adapters.RecyclerViewAdapter

/**
 * Created by RKR on 11-09-2018.
 * MultiSelectDialogFragment.
 */
open class MultiSelectDialogFragment<T : Selectable> : BaseDialogFragment() {

    private var mRequestCode: Int = 0
    private var mMultiSelectable = true
    private var mSelectedItem: Selectable? = null

    private var mAdapter: RecyclerViewAdapter? = null

    private lateinit var mListener: (List<Selectable>?) -> (Unit)

    /**
     * gets the adapter
     *
     * @return RecyclerViewAdapter
     */
    val adapter: RecyclerViewAdapter
        get() {
            if (mAdapter == null) {
                mAdapter = RecyclerViewAdapter()
                mAdapter?.setOnItemClickListener { `object`: Any?, _: Int, _: View ->
                    onClickItem(`object`)
                }
            }
            return mAdapter as RecyclerViewAdapter
        }

    @Suppress("UNCHECKED_CAST")
    private val selectedItems: List<Selectable>?
        get() {
            return if (mMultiSelectable) {
                adapter.data?.filterIsInstance(Selectable::class.java)
                    ?.filter { it.selected == true }
            } else mSelectedItem?.let { listOf(it) }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mRequestCode = arguments?.getInt(KEY_REQUEST_CODE) ?: 0
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_multi_select, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val recyclerView = view.findViewById<RecyclerView>(R.id.recycler_view)
        recyclerView?.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(context)
        recyclerView?.adapter = adapter
        view.findViewById<View>(R.id.action_positive)?.setOnClickListener { actionPositive() }
        view.findViewById<View>(R.id.action_negative)?.setOnClickListener { dismiss() }
    }

    @Suppress("unused")
    fun setMultiSelectable(multiSelectable: Boolean) {
        this.mMultiSelectable = multiSelectable
    }

    @Suppress("unused")
    fun setData(data: List<Selectable>?) {
        adapter.setTypedData(`data`?.takeIf { it.isNotEmpty() }?.apply {
            val mutableData = this as MutableList
            val selectedItems = selectedItems
            selectedItems?.forEach {
                if (mutableData.contains(it)) {
                    val index = mutableData.indexOf(it)
                    mutableData[index].selected = it.selected
                } else {
                    mutableData.add(0, it)
                }
            }
        })
    }

    private fun onClickItem(`object`: Any?) {
        if (!mMultiSelectable) {
            adapter.data?.filterIsInstance(Selectable::class.java)
                ?.find { it.selected == true }?.selected = false
            mSelectedItem = `object` as Selectable
            `object`.selected = true
            actionPositive()
        }
    }

    private fun actionPositive() {
        mListener(selectedItems)
        dismiss()
    }

    fun setListener(listener: (List<Selectable>?) -> (Unit)) {
        mListener = listener
    }

    companion object {

        private const val KEY_REQUEST_CODE = "REQUEST_CODE"

        fun <T : Selectable> newInstance(
            requestCode: Int,
            listener: (List<Selectable>?) -> (Unit)
        ): MultiSelectDialogFragment<T> {

            val args = Bundle()
            args.putInt(KEY_REQUEST_CODE, requestCode)
            val fragment = MultiSelectDialogFragment<T>()
            fragment.arguments = args
            fragment.isCancelable = false
            fragment.mListener = listener
            return fragment
        }
    }

}
